### ISSUE TITLE
Loops not disposing IEnumerator object returned by IEnumerable

### ISSUE THREAD

**Issue by jborean93** [PowerShell/PowerShell#20854]
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)
- [X] Search the existing issues.
- [X] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).
- [X] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://learn.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).

### Steps to reproduce

If an `IEnumerable` is looped the `IEnumerator` that it uses should be disposed if it implements `IDisposable`. If using the formatter which enumerates the `IEnumerable` the `IEnumerator` is disposed but using a pipe or `foreach` loop will never dispose the `IEnumerator`. This is bad if the `IEnumerator` holds some resources that needs to be disposed like a file handle or memory that should be cleaned up.

A reproducer for this

```powershell
class MyEnumerator : System.Collections.IEnumerator, IDisposable
{
    hidden [int]$count = 0

    [object] get_Current()
    {
        [Console]::WriteLine("get_Current()")
        return $this.count
    }
    
    [bool] MoveNext() {
        [Console]::WriteLine("MoveNext()")
        $this.count += 1
        return $this.count -lt 4
    }
    
    Reset() {
        [Console]::WriteLine("Reset()")
        $this.count = 0
    }

    Dispose() {
        [Console]::WriteLine("Dispose()")
    }
}

class MyEnumerable : System.Collections.IEnumerable {
    MyEnumerable() {}
    
    [System.Collections.IEnumerator] GetEnumerator() {
        [Console]::WriteLine("GetEnumerator()")
        return [MyEnumerator]::new()
    }
}

# Dispose is called
[MyEnumerable]::new()

# Dispose is not called for the following
# Any pipeline cmdlet can be used here
[MyEnumerable]::new() | ForEach-Object { $_ }

foreach ($obj in [MyEnumerable]::new()) {
    $obj
}
```

I can see this behaviour in WinPS 5.1 as well.

### Expected behavior

```console
All instances where PowerShell gets the `IEnumerator` object from the `IEnumerable` (it is the owner and thus should dispose once done) are disposed.
```

### Actual behavior

```console
Any pipelining or `foreach` looping will never have `Dispose()` be called.
```

### Error details

```console
None
```

### Environment data

```powershell
Name                           Value
----                           -----
PSVersion                      7.4.0
PSEdition                      Core
GitCommitId                    7.4.0
OS                             Arch Linux
Platform                       Unix
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

**Comment by jborean93** [comment]
Same behaviour if using an actual compiled assembly in C#

```powershell
Add-Type -TypeDefinition @'
using System;
using System.Collections;

public class MyEnumerator : IEnumerator, IDisposable
{
    private int count = 0;

    public object Current
    {
        get {
            Console.WriteLine("get_Current()");
            return count;
        }
    }

    public bool MoveNext()
    {
        Console.WriteLine("MoveNext()");
        count += 1;
        return count < 4;
    }
    
    public void Reset()
    {
        Console.WriteLine("Reset()");
        count = 0;
    }

    public void Dispose()
    {
        Console.WriteLine("Dispose()");
    }
}

public class MyEnumerable : IEnumerable
{
    public System.Collections.IEnumerator GetEnumerator()
    {
        Console.WriteLine("GetEnumerator()");
        return new MyEnumerator();
    }
}
'@
```

An annoying workaround is to get the `IEnumerator` yourself and dispose of it

```powershell
$enumerator = [MyEnumerable]::new().GetEnumerator()
$enumerator | ForEach-Object { $_ }
$enumerator.Dispose()

$enumerator = [MyEnumerable]::new().GetEnumerator()
foreach ($obj in $enumerator) { $obj }
$enumerator.Dispose()
```

**Comment by rhubarb-geek-nz** [comment]
I do not think it is part of the contract of [IEnumerable.GetEnumerator Method](https://learn.microsoft.com/en-us/dotnet/api/system.collections.ienumerable.getenumerator?view=net-8.0) that you have to check to see if it is [IDIsposable](https://learn.microsoft.com/en-us/dotnet/api/system.idisposable?view=net-8.0)

If it was, they would have mentioned it.

If it was part of the contract then the interface [IEnumerator](https://learn.microsoft.com/en-us/dotnet/api/system.collections.ienumerator?view=net-8.0) would inherit from [IDisposable](https://learn.microsoft.com/en-us/dotnet/api/system.idisposable?view=net-8.0)

**Comment by jborean93** [comment]
> I do not think it is part of the contract of [IEnumerable.GetEnumerator Method](https://learn.microsoft.com/en-us/dotnet/api/system.collections.ienumerable.getenumerator?view=net-8.0) that you have to check to see if it is [IDIsposable](https://learn.microsoft.com/en-us/dotnet/api/system.idisposable?view=net-8.0)

C# has automatically done a check on `IDisposable` on the enumerator since C# 1.2. You can see how it actually compiles a `foreach` loop the code with https://sharplab.io/#v2:D4AQTAjAsAUCAMACEEB0BhA9gG2wUwGMAXAS0wDsBnAbllhAGZkxF1EBvWRb5JkAFkQBZABQBKDlx7SAZpgBOeAIYEAFohEB5AEYArQkURzMiEuUTk8Ad2EBPAKLkArgFs88pdvzixU6d04Yf2DkCABOEWMxWiCQgF8/RASYZPo+FiEHZzcPIgVEAC5EAElHV3clPPkAGhKAERJKAAdMSk98WEDpJvkSADdKvFNyQwJMJxHEAF5EeBjExkRMPQNWJ3lFEcSukIBzPCJE6R2QnhQIgCJ9ogB9dHXNonEL6KOQkAB2RDGJohjT5LSVKxbiLbSYHDCTB9PAAOTwAA8nr4QZJUdJziILkJoXDEciXv8Qj9JgBqGYQInBT7fcaTAA8iH4VKSiQWfEEACU8JQDuJtm9QeEsdzeQTXuieCTDDM5olgRiOYgGs1Wnh+aiTv5MRcVS1ec8JUDYMDFuA7GUcu0hkUUAxOuzQhgcPhiGQqKhStkKlVEABxA6Wn0KDXHQWhS4BohB3IhsSE8M0yw2TIxypxlnJOJAA==

Notice how it transforms the `foreach` loop into a `while` loop that enumerates the `IEnumerator` and does an `IDisposable` check at the end. PowerShell should be doing the same thing here anytime it is enumerating an `IEnumerator` that itself created, i.e. when calling `GetEnumerator()` on an `IEnumerable`.

PowerShell not doing so is a change in behaviour and is problematic if the `IEnumerator` has any unmanaged resources that should be disposed when no longer in use.

**Comment by rhubarb-geek-nz** [comment]
While code generated by foreach may be doing that out of politeness I still say it is not part of the contract. If it was then the Microsoft documentation would have mentioned it.

> PowerShell not doing so is a change in behaviour and is problematic if the `IEnumerator` has any unmanaged resources that should be disposed when no longer in use.

If the solution in C# is to use foreach then I suggest that the solution in PowerShell is also to use [foreach](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_foreach?view=powershell-7.4)/[ForEach-Object](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/foreach-object?view=powershell-7.4) rather than calling GetEnumerator() yourself.

**Comment by jborean93** [comment]
> While code generated by foreach may be doing that out of politeness I still say it is not part of the contract. If it was then the Microsoft documentation would have mentioned it.

Look at the [ECMA-334 spec](https://www.ecma-international.org/wp-content/uploads/ECMA-334_6th_edition_june_2022.pdf) for `foreach` loop under section `12.9.5 The foreach statement` specifically how it talks about the `finally` block for this statement. This is the definitive spec for the C# language and I would strongly argue that PowerShell would follow a lot of the conventions around this.

This has been the behaviour of C# since 2003 https://learn.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-version-history#c-version-12. The fact that PowerShell does not do this means that if you are trying to enumerate an `IEnumerable` that returns a disposable `IEnumerator` using `foreach (...)` or through the pipeline it may leave behind unmanaged resources which are never cleaned up. Granted this may not be a common scenario but it's still very surprising to see this behaviour.

> If the solution in C# is to use [foreach](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_foreach?view=powershell-7.4) then I suggest that the solution in PowerShell is also to use foreach/[ForEach-Object](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/foreach-object?view=powershell-7.4) rather than calling GetEnumerator() yourself.

That's my point, PowerShell does not do this while C# does for its loops. To do this yourself you need to get the `IEnumerator` and dispose of it yourself

```powershell
$enumerator = [MyEnumerable]::new().GetEnumerator()
try {
    foreach ($obj in $enumerator) { $obj }
}
finally {
    if ($enumerator -is [IDisposable]) { $enumerator.Dispose() }
}
```

**Comment by rhubarb-geek-nz** [comment]
I had a quick look in the code

```
$ find . -name "*.cs" | xargs grep GetEnumerator | wc -l
195
```

Of special interest is WriteToPipelineWithUnrolling in src/System.Management.Automation/engine/InternalCommands.cs
